### PR TITLE
.github: Enable SSH to linux runners

### PR DESCRIPTION
**Pull request by seemethere**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62280

Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Depends on https://github.com/fairinternal/pytorch-gha-infra/pull/8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Closes https://github.com/fairinternal/pytorch-gha-infra/issues/5

Differential Revision: [D29941681](https://our.internmc.facebook.com/intern/diff/D29941681)

### CODE REVIEW

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29941681).

File: .github/templates/linux_ci_workflow.yml.j2 [hunk 1]
```diff
@@ -231,6 +237,19 @@ jobs:
           path:
             artifacts.zip
       {%- endif %}
+      - name: Hold runner for 2 hours or until ssh sessions have drained
+        # Always hold for active ssh sessions
+        if: always()
+        run: |
+          echo "Holding runner for 2 hours until all ssh sessions have logged out"
```
Comment (janeyx99) [inline .github/templates/linux_ci_workflow.yml.j2:244]: This means that even if the ssh session is on, the runner will still shut down after 2 hours regardless, right?

Comment (seemethere) [reply .github/templates/linux_ci_workflow.yml.j2:244]: Yes, this is consistent with CircleCI's current re-run with ssh functionality, see: https://circleci.com/docs/2.0/ssh-access-jobs/

File: .github/templates/linux_ci_workflow.yml.j2 [hunk 2]
```diff
@@ -564,6 +606,19 @@ jobs:
...
+          for _ in $(seq 1440); do
+            # Break if no ssh session exists anymore
+            if [[ "$(who)" = "" ]]; then
+              break
+            fi
+            echo "."
+            sleep 5
+          done
```
Comment (samestep) [inline .github/templates/linux_ci_workflow.yml.j2:621]: looks like this same `run` script is used in a few different places; would it make sense to put it into `.github/scripts` and just call it from here?

**Review by samestep — APPROVED** [review]
lgtm! I guess most of the logic is in https://github.com/seemethere/add-github-ssh-key?

**Comment by seemethere** [comment]
> lgtm! I guess most of the logic is in [seemethere/add-github-ssh-key](https://github.com/seemethere/add-github-ssh-key)?

Yup most of the logic for this particular feature is in the terraform for the security groups as well as the the github actions to add the ssh keys.

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29941681).

**Review by janeyx99 — APPROVED** [review]

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29941681).

**Comment by JpD9119** [comment]
[เปิดใช้งาน SSH เพื่อรันลินุกซ์ GHA](https://github.com/pytorch/pytorch.wiki.git) 

> ``

**Pull request closed without merging**